Data/Binary/Serialise/CBOR/Read.hs

```diff
@@ -1882,7 +1889,19 @@ readBigNInt bs
 nintegerFromBytes :: BS.ByteString -> Integer
 nintegerFromBytes bs = -1 - uintegerFromBytes bs
 
--- TODO FIXME: optimised implementation for GMP
+#if defined(OPTIMIZE_GMP)
+uintegerFromBytes :: BS.ByteString -> Integer
+uintegerFromBytes (BS.PS fp (I# off#) (I# len#)) =
+  -- This should be safe since we're simply reading from ByteString (which is
+  -- immutable) and GMP allocates a new memory for the Integer, i.e., there is
+  -- no mutation involved.
+  unsafePerformIO $
+      withForeignPtr fp $ \(Ptr addr#) ->
+          let addrOff# = addr# `plusAddr#` off#
+          -- The last parmaeter (`1#`) tells the import function to use big
+          -- endian encoding.
+          in Gmp.importIntegerFromAddr addrOff# (int2Word# len#) 1#
```

Comment:
We could actually follow the same trick as for writing and avoid the `-1-n` conversion by going straight to `importBigNatFromAddr` and using `Gmp.Jn#`.